### DISCUSSION TITLE
Type declarations for notNil can be tricked into believing that null is, in fact, not nil

### DISCUSSION THREAD

**Discussion by kirkwaiblinger** [iter-tools/iter-tools#489] [General]
This is a crazy nitpick, but I noticed something interesting while playing with this library (which, I would like to emphasize I am a fan of).

It seems that the `notNil` function can be fooled by manually specifying the generic type to be nullable. I made a [TS Playground example](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtX1RwwDksIAeAFQD4AKADwC55L4AfA5CCd+NURFlQhgASmb14WAM4sA3ACgFoSLAQp02PASLEuEGQAsqdJi3HxJM+MTx7uUAEYQQJxcvDQ4SNJlz5CEn0jShwcEwZmVH1efhBBYTFmRCgIaRBFFS91Xy0A3WDpQ1DwmkjzCSlZW1R7aGdXGncAemb4AFkoIT50+DAoXoBJeABrQgB3eBxEJBwYAhARIQBzHRIyJSy1PrxpDHhYGCgATwB5RAAFHGlpLGdjusGMEABbaWYAQRgj4-I9mBWvGi3Go7gA3gp4FCdqg9vAijguMAAEIgD43LDLVBOFyhL4-c41OpGT7fE5-DAA1DLajwAC8BzJZ0u11u90ezzeADpBBBnjBaAxRPTaZIAIQM4EQUQKAC+Cgh0JhcIRSNR6NuWJxIDxTMJdkKhlJPwpVJp9MZBJZGPZ+ier2kPLI-NogVI0rlCsh0LAu32qogKLRGK1DV1VqJhuN5P+K1pDMOJ3OVxtEAeds5jt5Lrdhpl8sVPr98MMiMD6pD2LDOHxScQke4JPgtd+sep8ctdZTbLTHIdTr5IAFucbRRK+aUClaLBAcIwhig+3GCH6qECsxwPHnCAwxwADghpMc9q8pKgMDh4POhCNAfPF1fDAhYMtkC8QOer5eRwYin0oPgexkDwjgIFK8DjE++RBKOhjwHAGDIDAsJXjAyAgFymyeNsvqwv6lIrKcMAAKqoAIQgiMwbarBwcQJCIShYDMtDUURpHkYk8ASpw3DCoWUK4SqpZqsGmpVriOAAMoEdSVEyasDKsSRZHxBRwCKPKChMfArq6GQpqArRKn0cAdBKexqmJKIfHegJxYBkGGqYuJOpSfJclmha5nGWpGmMcxP5GAZ1KxD5iRmfJbFhSI1nwPxyr+sJ5aic52qhNJZoeYCimRcpHEiH5WkBQUsElMFNF8NFpksblFkmbF8WCYlZaOZWaVuZl8LyV5tVVYVU5tJQs77Pe+zILcIVQKMEwBIa8AAG6pOhEHAfBICIchBxIKkvQGPyqQ6cIIgiPAjjHPAAACu4HtIYAAnuGAALSzgY57NIQj1oMIYCztIsDHI9uHAFgeQyh4qjeE18AcSDIC9lGPE8EZ+XqeD2QJdDlmw72JW-kaXVmuCtkYw5ABK61IagABiO0gFTczEkUyS0xaMPPPDsFcZK+h+Y19lJcA5MbdTtP0zAjP4ykaQIJKekQLQbNw+msETnzeEli1QuUzT0tixLzPSxagVFArWPs8reOq8TUNkxTyE6+kesI1LvSyzBeMlKbCTYxbRgTgNM5zgu+xTWMODjLNnOLRAy3jKtoEEIc4cnReCzzUOmFFTpisc3jXOIzZSo2wLqLECA6cwFE5dDqzZtKxLnpZ7p6zyznvtFA11v8y1pfV5Xac1wybcN5p2nNxLXtCD7Eud0X3ciWXFdVxXtfe+bI-+YduMhGEk+oNPeaF0W6sOb3S8D-MQ917nRiN9OwzAHgADk+xh5MpaTKnLxQCMz7cFMMxty9D3AMR0CggA) to elaborate a bit.

Not very important but thought you might find it interesting. Would be happy to submit a PR if deemed beneficial after discussion. Also, I'm very curious is someone does know how to write declarations that pass each of the test cases that I wrote, since I could not figure out a way to do so. I could only figure out how to pass more than the current declarations (though, of course, my suggestions may fail for test cases I am unaware of).

**Comment by conartist6** [comment]
Huh. I can confirm that no, I never even considered that possibility. It doesn't look to me like you can get all those cases right with a single typedef either.

**Reply by kirkwaiblinger** [reply]
Fair, yeah, I really can't conceive of a case in which anyone would even accidentally trigger the faulty case with isNil. The most likely way (I guess?) that someone could trigger it could be with `notNil<typeof x>(x)`.... but again this seems like a bizarre thing to do. Also TIL `array.filter(notNil<SomeType | null>)` is valid TS to specify the generic. Again, not sure how you could find yourself needing those cases, but it appears the `NonNullable<T>` version of the type declarations would permit it.

**Reply by conartist6** [reply]
I didn't know that about generics...

**Reply by kirkwaiblinger** [reply]
As said, it's niche, to be sure, but I know of several scenarios where specifying the generic on a function call is concretely useful in application code. Here are two
1. querying DOM. `const aCanvasElement = document.querySelector<HTMLCanvasElement>('#myCustomIdThatDefinitelyIsACanvas')`. Then `aCanvasElement` will have type `HTMLCanvasElement | null`, rather than `Element | null`. (Ok, this is kind of a bad example because this is no safer than a type assertion. There is a similar example with axios requests - but again, those are basically type assertions)
2. VueJS typing an uninitialized ref. `const maybeString = ref<string | undefined>(undefined)`. Without specifying the generic, `maybeString` would have type `Ref<undefined>` which wouldn't be very useful. This one is actually typesafe. Note that `const maybeString: Ref<string | undefined> = ref(undefined)` would work too but it requires types to be specified on the LHS, rather than coming entirely from the RHS definition.

**Reply by conartist6** [reply]
I'd just write `const maybeString = ref(undefined as string | undefined)`

**Reply by kirkwaiblinger** [reply]
True, true. That would definitely work too. Though some sort of extreme pedant such as myself might complain of not being able to [ban type assertions altogether](https://typescript-eslint.io/rules/consistent-type-assertions/#assertionstyle) with that pattern.

Anyhow, if I'm making a point at all, my point is just that I think my suggestion could make the `notNil` slightly more pedantically strict (via the declarations for the function I called `notNullish` in the TS playground), if you were interested in using that in this library. That's all :)

**Reply by conartist6** [reply]
If it's important to you to ban type assertions you totally can, you just have to use a method to construct your widened primitives...

**Reply by conartist6** [reply]
Since iter-tools is relatively mature I tend to do work on the project in complete passes. At the moment I'd say I do not support passing explicit types to function parameters. Obviously you can do it, and maybe some people do, but I don't support it. I think the change you really want is for the library to start supporting the pattern, and that's a big process that involves auditing the code to find the risks and bad patterns, documenting what can and can't be relied on to ensure that semver continues to protect users from breakages, and of course testing to ensure that what works stays working over time.

My focus is elsewhere so I won't be driving this. I'm more likely to solve the problem by saying explicitly in the documentation that you should never directly specify type parameters, and I'm not responsible for what happens when you do.

That being the case, how do you feel about the options? Are you interested in tackling the work to get the feature to supported, bearing in mind that you'd first have to convince me that it could be supported reasonably well.

**Reply by conartist6** [reply]
You'd also have to convince me that there is a real problem, which I am not convinced of yet. I can see banning narrowing assertions (they are unsafe), but I don't see the logic to banning widening assertions, which are safe. If tooling can't tell the difference then I'm pretty sure that's a problem with the tooling.
